analytic_engine/src/sst/parquet/encoding.rs

```diff
@@ -461,3 +487,457 @@ impl ParquetEncoder {
         self.record_encoder.close()
     }
 }
+
+/// RecordDecoder is used for decoding ArrowRecordBatch based on
+/// `schema.StorageFormat`
+trait RecordDecoder {
+    fn decode(&self, arrow_record_batch: ArrowRecordBatch) -> Result<ArrowRecordBatch>;
+}
+
+struct ColumnarRecordDecoder {}
+
+impl RecordDecoder for ColumnarRecordDecoder {
+    fn decode(&self, arrow_record_batch: ArrowRecordBatch) -> Result<ArrowRecordBatch> {
+        Ok(arrow_record_batch)
+    }
+}
+
+struct HybridRecordDecoder {}
+
+impl HybridRecordDecoder {
+    /// Convert `ListArray` fields to underlying data type
+    fn convert_schema(arrow_schema: ArrowSchemaRef) -> ArrowSchemaRef {
+        let new_fields: Vec<_> = arrow_schema
+            .fields()
+            .iter()
+            .map(|f| {
+                if let DataType::List(nested_field) = f.data_type() {
+                    Field::new(f.name(), nested_field.data_type().clone(), true)
+                } else {
+                    f.clone()
+                }
+            })
+            .collect();
+        Arc::new(ArrowSchema::new_with_metadata(
+            new_fields,
+            arrow_schema.metadata().clone(),
+        ))
+    }
+
+    /// Stretch hybrid collpased column into columnar column.
+    /// `value_offsets` specify offsets each value occupied, which means that
+    /// the number of a `value[n]` is `value_offsets[n] - value_offsets[n-1]`.
+    /// Ex:
+    ///
+    /// `array_ref` is `a b c`, `value_offsets` is `[0, 3, 5, 6]`, then
+    /// output array is `a a a b b c`
+    ///
+    /// Note: caller should ensure offsets is not empty.
+    fn stretch_variable_length_column(
+        array_ref: &ArrayRef,
+        value_offsets: &[i32],
+    ) -> Result<ArrayRef> {
+        assert!(!value_offsets.is_empty());
+
+        let values_num = *value_offsets.last().unwrap() as usize;
+        let offset_slices = array_ref.data().buffers()[0].as_slice();
+        let value_slices = array_ref.data().buffers()[1].as_slice();
+        let null_bitmap = array_ref.data().null_bitmap();
+        debug!(
+            "raw buffer slice, offsets:{:#02x?}, values:{:#02x?}, bitmap:{:#02x?}",
+            offset_slices,
+            value_slices,
+            null_bitmap.map(|v| v.buffer_ref().as_slice())
+        );
+
+        let mut i32_offsets = Vec::with_capacity(offset_slices.len() / OFFSET_SIZE);
+        for i in (0..offset_slices.len()).step_by(OFFSET_SIZE) {
+            let offset = i32::from_le_bytes(offset_slices[i..i + OFFSET_SIZE].try_into().unwrap());
+            i32_offsets.push(offset);
+        }
```

Comment:
We have a similar logic in `get_list_array_offsets()`. How about making `get_list_array_offsets` accept a slice and reuse it in both places?